ecs-agent/acs/session/session.go

```diff
@@ -11,13 +11,348 @@
+	// and make sure Agent does not force ACS to send credentials for any subsequent reconnects to ACS.
+	logger.Info("Connected to ACS endpoint")
+
+	s.sendCredentials = false
+
+	return s.startACSSession(ctx, client)
+}
+
+// startACSSession starts a session with ACS. It adds request handlers for various
+// kinds of messages expected from ACS. It returns on server disconnection or when
+// the context is canceled.
+func (s *session) startACSSession(ctx context.Context, client wsclient.ClientServer) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	responseSender := func(response interface{}) error {
+		return client.MakeRequest(response)
+	}
+	responders := []wsclient.RequestResponder{
+		NewPayloadResponder(s.payloadMessageHandler, responseSender),
+		NewRefreshCredentialsResponder(s.credentialsManager, s.credentialsMetadataSetter, s.metricsFactory,
+			responseSender),
+		NewAttachTaskENIResponder(s.eniHandler, responseSender),
+		NewAttachInstanceENIResponder(s.eniHandler, responseSender),
+		NewHeartbeatResponder(s.doctor, responseSender),
+		NewTaskManifestResponder(s.taskComparer, s.sequenceNumberAccessor, s.manifestMessageIDAccessor,
+			s.metricsFactory, responseSender),
+		NewTaskStopVerificationACKResponder(s.taskStopper, s.manifestMessageIDAccessor, s.metricsFactory),
+	}
+	for _, r := range responders {
+		client.AddRequestHandler(r.HandlerFunc())
+	}
+
+	if s.dockerVersion == "containerd" && s.resourceHandler != nil {
+		client.AddRequestHandler(NewAttachResourceResponder(s.resourceHandler, s.metricsFactory,
+			responseSender).HandlerFunc())
+	}
+
+	if s.dockerVersion != "containerd" && s.addUpdateRequestHandlers != nil {
+		s.addUpdateRequestHandlers(client)
+	}
+
+	// Start a heartbeat timer for closing the connection.
+	heartbeatTimer := newHeartbeatTimer(client, s.heartbeatTimeout, s.heartbeatJitter)
+	// Any message from the server resets the heartbeat timer.
+	client.SetAnyRequestHandler(anyMessageHandler(heartbeatTimer, client))
+	defer heartbeatTimer.Stop()
+
+	backoffResetTimer := time.AfterFunc(
+		retry.AddJitter(s.heartbeatTimeout, s.heartbeatJitter), func() {
+			// If we do not have an error connecting and remain connected for at
+			// least 1 or so minutes, reset the backoff. This prevents disconnect
+			// errors that only happen infrequently from damaging the reconnect
+			// delay as significantly.
+			s.backoff.Reset()
+		})
+	defer backoffResetTimer.Stop()
+
+	return client.Serve(ctx)
+}
+
+func (s *session) reconnectDelay(acsError error) (time.Duration, bool) {
+	if isInactiveInstanceError(acsError) {
+		logger.Info("Container instance is deregistered")
+		s.inactiveInstanceCB()
+		return s.inactiveInstanceReconnectDelay, true
+	}
+	if shouldReconnectWithoutBackoff(acsError) {
+		// ACS has closed the connection for valid reasons. Example: periodic disconnect.
+		// No need to wait/backoff to reconnect.
+		logger.Info("ACS WebSocket connection closed for a valid reason")
```

Comment:
I used level INFO since that is currently being used for this scenario. See current corresponding code here:
https://github.com/aws/amazon-ecs-agent/blob/dev/agent/acs/handler/acs_handler.go#L180